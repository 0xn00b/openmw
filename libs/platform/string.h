// Wrapper for string.h on Mac
#ifndef _STRING_WRAPPER_H
#define _STRING_WRAPPER_H

#include <string.h>
#ifdef __APPLE__
// need our own implementation of strnlen
static size_t strnlen(const char *s, size_t n)
{
  const char *p = (const char *)memchr(s, 0, n);
  return(p ? p-s : n);
}
#endif
#endif
