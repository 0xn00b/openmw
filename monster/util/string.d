/*
  Monster - an advanced game scripting language
  Copyright (C) 2004, 2007, 2008  Nicolay Korslund
  Email: <korslund@gmail.com>
  WWW: http://monster.snaptoad.com/

  This file (string.d) is part of the Monster script language
  package.

  Monster is distributed as free software: you can redistribute it
  and/or modify it under the terms of the GNU General Public License
  version 3, as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  version 3 along with this program. If not, see
  http://www.gnu.org/licenses/ .

 */

module monster.util.string;

import std.utf;
import std.string;

// These functions check whether a string begins or ends with a
// certain substring.
bool begins(char[] str, char[] start)
{
  if(str.length < start.length ||
     str[0..start.length] != start) return false;
  return true;
}

bool ends(char[] str, char[] end)
{
  if(str.length < end.length ||
     str[$-end.length..$] != end) return false;
  return true;
}

// Case insensitive versions of begins and ends
bool iBegins(char[] str, char[] start)
{
  if(str.length < start.length ||
     icmp(str[0..start.length], start) != 0) return false;
  return true;
}

bool iEnds(char[] str, char[] end)
{
  if(str.length < end.length ||
     icmp(str[$-end.length..$], end) != 0) return false;
  return true;
}

char[] nextWord(ref char[] str, char delim = ' ')
{
  int i = find(str, delim);
  char[] result;

  // No 'delim' found, return the entire string and set remainder to
  // null.
  if(i == -1)
    {
      result = str;
      str = null;
      return result;
    }

  // A separator was found. Return everything upto 'delim' (index i),
  // put the remainder of the string (not including the char at [i])
  // in str.
  result = str[0..i];
  str = str[i+1..$];
  return result;
}

// Strip trailing zeros
char[] stripz(char [] s)
{
  foreach(int i, char c; s)
    if( c == 0 )
      return s[0..i];

  return s;
}

// Convert a long integer into a string using nice comma
// formatting. delim is the delimiter character, size is the number of
// digits in each group. See the unittest for examples.
char[] comma(long i, char delim=',', int size = 3)
{
  char[] str = toString(i);
  char[] res;

  if(i<0) str=str[1..$];

  str.reverse;
  foreach(int j, char c; str)
    { 
      if(j!=0 && j%size == 0)
        res = delim ~ res;
      res = c ~ res;
    }

  if(i<0) res = "-" ~ res;

  return res;
}
