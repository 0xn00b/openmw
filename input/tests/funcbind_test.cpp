#include <iostream>
using namespace std;

#include <string>
#include <vector>
#include <boost/function.hpp>
#include <assert.h>

typedef boost::function<void()> Action;

struct FuncBind
{
  std::string name;
  Action action;
};

class Binder
{
  std::vector<FuncBind> bindings;

public:
  /**
     Initialize the struct by telling it how many functions you have
     to bind. The largest index you intend to use should be number-1.
  */
  Binder(int number) : bindings(number) {}

  void bind(int index, Action action, const std::string &name="")
  {
    assert(index >= 0 && index < bindings.size());
    FuncBind &fb = bindings[index];
    fb.action = action;
    fb.name = name;
  }

  void unbind(int index)
  {
    assert(index >= 0 && index < bindings.size());
    FuncBind &fb = bindings[index];
    fb = FuncBind();
  }

  void call(int index)
  {
    assert(index >= 0 && index < bindings.size());
    FuncBind &fb = bindings[index];
    if(fb.action)
      {
        cout << "Calling '" << fb.name << "'\n";
        fb.action();
      }
    else
      cout << "No function\n";
  }
};

void f1()
{
  cout << "In f1()\n";
}

void f2()
{
  cout << "In f2()\n";
}

int main()
{
  cout << "This will test the function binding system\n";

  Binder bnd(5);

  bnd.bind(0, &f1, "This is action 1");
  bnd.bind(1, &f2);
  bnd.bind(2, &f1, "This is action 3");

  for(int i=0; i<5; i++)
    {
      cout << "\nCalling " << i << endl;
      bnd.call(i);
    }

  return 0;
}
