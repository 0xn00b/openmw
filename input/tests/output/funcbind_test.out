This will test the function binding system

Calling 0
Calling 'This is action 1'
In f1()

Calling 1
Calling ''
In f2()

Calling 2
Calling 'This is action 3'
In f1()

Calling 3
No function

Calling 4
No function
